esi_leap/db/sqlalchemy/api.py

```diff
@@ -133,23 +133,139 @@ def offer_get_all(filters):
+                a.pop(i)
+            else:
+                i += 1
+
+        a = [[a[j], a[j + 1]] for j in range(0, len(a) - 1, 2)]
+
+    else:
+        a = [[offer_ref.start_time, offer_ref.end_time]]
+
+    return a
+
+
+def offer_verify_availability(offer_ref, start_time=None, end_time=None):
+    availabilities = offer_get_availabilities(offer_ref)
+
+    if availabilities == []:
+        raise exception.OfferNotAvailable(offer_uuid=offer_ref.uuid,
+                                          start_time=start_time,
+                                          end_time=end_time)
+
+    start = start_time if start_time else None
+    end = end_time if end_time else None
+
+    available = False
+
+    l = 0
+    h = len(availabilities) - 1
+    m = (l + h) // 2
+
+    if start is None and end:
+
+        if end > availabilities[-1][1]:
+            raise exception.OfferNotAvailable(offer_uuid=offer_ref.uuid,
+                                              start_time=start_time,
+                                              end_time=end_time)
+        while l < h:
+            if availabilities[m][1] <= end:
+                l = m + 1
+            else:
+                h = m
+            m = (l + h) // 2
+
+        start_time = availabilities[m][0]
+
+    elif start:
+
+        if start < availabilities[0][0]:
+            raise exception.OfferNotAvailable(offer_uuid=offer_ref.uuid,
+                                              start_time=start_time,
+                                              end_time=end_time)
+        while l <= h:
+            if availabilities[m][0] > start:
+                h = m - 1
+            else:
+                l = m + 1
+            m = (l + h) // 2
+
+        if end is None:
+            end_time = availabilities[m][1]
+    else:
+        m = 0
+        start_time = availabilities[0][0]
+        end_time = availabilities[0][1]
+
+    if start_time >= end_time:
+        raise exception.InvalidTimeRange(resource="a contract",
+                                         start_time=str(start_time),
+                                         end_time=str(end_time))
+
+    if start_time >= availabilities[m][0] and end_time <= availabilities[m][1]:
+        available = True
+
+    if not available:
+        raise exception.OfferNotAvailable(offer_uuid=offer_ref.uuid,
+                                          start_time=start_time,
+                                          end_time=end_time)
+
+    return start_time, end_time
+
+
 def offer_create(values):
 
     offer_ref = models.Offer()
     values['uuid'] = uuidutils.generate_uuid()
 
-    if 'start_time' in values and 'end_time' in values and \
-       values['start_time'] is not None and values['end_time'] is not None and \
-       values['start_time'] >= values['end_time']:
+    if 'start_time' not in values:
+        values['start_time'] = datetime.datetime.now()
+    if 'end_time' not in values:
+        values['end_time'] = datetime.datetime.max
```

Comment:
If we're defaulting `project_id` at the REST API level, then we should be defaulting these values there as well.

Reply:
I did this because defaulting project id is a policy protection issue and what happens to it varies based on policy, but here the times are a data integrity issue. The whole thing just won't work if an end_time is set to null

Reply:
Isn't there a database constraint that would also prevent that? If not, then we should just add 'nullable=false' to the appropriate fields in the database model.